src/core/socket.c

```diff
@@ -2039,13 +2089,29 @@ static void socket_enter_running(Socket *s, int cfd) {
         } else {
                 _cleanup_free_ char *prefix = NULL, *instance = NULL, *name = NULL;
                 Service *service;
+                SocketPeer *p;
 
                 if (s->n_connections >= s->max_connections) {
                         log_unit_warning(UNIT(s), "Too many incoming connections (%u), refusing connection attempt.", s->n_connections);
                         safe_close(cfd);
                         return;
                 }
 
+                if (s->max_connections_per_source > 0) {
+
+                        r = socket_find_peer(s, cfd, &p);
+                        if (r < 0) {
+                                safe_close(cfd);
+                                return;
+                        }
+
+                        if (p->n_ref > s->max_connections_per_source) {
+                                log_unit_warning(UNIT(s), "Too many incoming connections (%u) from source, refusing connection attempt.", p->n_ref);
```

Comment:
Shouldn't this be ratelimited? It is under external control, so if the other party know a service has this set they can easily saturate logs with minimal effort.

The same applies to `s->n_connections >= s->max_connections` above. Adding ratelimiting to both might be a good idea for another PR.